File: android/Gutenberg/src/main/java/org/wordpress/gutenberg/http/RequestBody.kt
package org.wordpress.gutenberg.http

import java.io.ByteArrayInputStream
import java.io.Closeable
import java.io.File
import java.io.InputStream
import java.io.RandomAccessFile

/**
 * An immutable, read-only view over a [ByteArray].
 *
 * This inline value class prevents callers from mutating the backing array
 * through the type system, without the overhead of a defensive copy.
 * At runtime the wrapper is erased — method calls compile to static
 * functions that operate directly on the underlying `ByteArray`.
 */
@JvmInline
value class ReadOnlyBytes(private val backing: ByteArray) {
    /** The number of bytes. */
    val size: Int get() = backing.size

    /** Returns the byte at the given [index]. */
    operator fun get(index: Int): Byte = backing[index]

    /** Returns a mutable copy of the underlying data. */
    fun copyOf(): ByteArray = backing.copyOf()

    /** Returns a mutable copy of the specified range. */
    fun copyOfRange(fromIndex: Int, toIndex: Int): ByteArray =
        backing.copyOfRange(fromIndex, toIndex)

    /** Creates an [InputStream] that reads from the backing data. */
    fun inputStream(): InputStream = ByteArrayInputStream(backing)

    /**
     * Finds the first occurrence of [pattern] starting from [fromIndex].
     * Returns -1 if not found.
     *
     * Uses a naive O(n·m) scan. This is sufficient because it is only used
     * for finding the header terminator (`\r\n\r\n`) within the first 64 KB
     * of a request — the pattern is 4 bytes and the data is capped by the
     * header size limit, so KMP/Boyer-Moore overhead is not justified.
     */
    fun indexOf(pattern: ByteArray, fromIndex: Int = 0): Int {
        if (pattern.isEmpty()) return fromIndex
        val limit = size - pattern.size
        if (fromIndex > limit) return -1
        outer@ for (i in fromIndex..limit) {
            for (j in pattern.indices) {
                if (backing[i + j] != pattern[j]) continue@outer
            }
            return i
        }
        return -1
    }

    /** Returns whether the bytes at [offset] match [pattern]. */
    fun startsWith(pattern: ByteArray, offset: Int): Boolean {
        if (offset + pattern.size > size) return false
        for (i in pattern.indices) {
            if (backing[offset + i] != pattern[i]) return false
        }
        return true
    }
}

/**
 * Owner for a temporary file, responsible for deleting it when no longer needed.
 *
 * Multiple [RequestBody] instances may share the same owner — for example, when
 * multipart parsing creates part bodies that reference byte ranges within the
 * same backing file. Because of this shared ownership, individual RequestBody
 * consumers should NOT close the owner; instead, the server's connection handler
 * calls [close] once the entire request (including all derived parts) is done.
 *
 * Active files are tracked in a companion registry so that orphaned temp files —
 * left behind by a crash or process kill — can be cleaned up on the next server
 * start via [cleanOrphans].
 *
 * ### Why not finalize() or deleteOnExit()?
 *
 * Both are unreliable on Android:
 * - `finalize()` is deprecated and ART may skip finalizer execution entirely
 *   under memory pressure (Android 12+).
 * - `deleteOnExit()` only fires on clean JVM shutdown. Android apps are killed
 *   by the OS, not shut down gracefully, so the hook effectively never runs.
 *
 * Instead, cleanup relies on two mechanisms:
 * 1. **Explicit [close]** at the connection boundary (primary path).
 * 2. **[cleanOrphans]** at server start (safety net for crashes).
 */
internal class TempFileOwner(val file: File) : Closeable {
    init {
        activeFiles.add(file.absolutePath)
    }

    override fun close() {
        activeFiles.remove(file.absolutePath)
        file.delete()
    }

    companion object {
        /** The default subdirectory name used for temp files under the injected cache dir. */
        const val DEFAULT_TEMP_SUBDIR = "gutenberg-http"

        /** Paths of temp files currently owned by a live [TempFileOwner]. */
        private val activeFiles = java.util.concurrent.ConcurrentHashMap.newKeySet<String>()

        /**
         * Deletes orphaned temp files in the [cacheDir]/[tempSubdir] directory.
         *
         * A file is considered orphaned if it is not tracked by any active
         * [TempFileOwner]. Call this at a safe point when no requests are
         * in-flight (e.g., server start).
         *
         * **Important:** Two server instances with the same `name` must not run
         * concurrently. On startup, this method deletes all files in the server's
         * temp subdirectory that are not tracked by an active [TempFileOwner]. If
         * another instance with the same name is still handling requests, its
         * in-flight temp files may be removed, causing I/O failures. Callers must
         * ensure each running server uses a unique name, or that the previous
         * instance is fully stopped before starting a new one.
         *
         * @param cacheDir The base cache directory.
         * @param tempSubdir The subdirectory name for temp files. When used via
         *   [HttpServer][org.wordpress.gutenberg.HttpServer], this is scoped by
         *   the server's `name` to prevent interference between concurrent instances.
         */
        fun cleanOrphans(cacheDir: File, tempSubdir: String = DEFAULT_TEMP_SUBDIR) {
            val dir = File(cacheDir, tempSubdir)
            if (!dir.isDirectory) return
            dir.listFiles()?.forEach { file ->
                if (!activeFiles.contains(file.absolutePath)) {
                    file.delete()
                }
            }
        }
    }
}

/**
 * An HTTP request body with stream semantics.
 *
 * `RequestBody` abstracts over the underlying storage (in-memory data or a file on disk)
 * and provides uniform access regardless of backing:
 * - **Stream access**: Use [inputStream] to read without loading everything into memory.
 * - **Materialized access**: Use [readBytes] to get the full contents.
 */
sealed class RequestBody {

    /** The number of bytes in the body. */
    abstract val size: Long

    /**
     * Creates an [InputStream] for reading the body contents.
     *
     * The caller is responsible for closing the returned stream.
     */
    abstract fun inputStream(): InputStream

    /** Reads the full body contents into a byte array. */
    abstract fun readBytes(): ByteArray

    /**
     * Reads the full body contents and returns a read-only view along with the
     * file offset at which the data begins (0 for in-memory bodies).
     *
     * This is used internally for multipart boundary scanning.
     */
    internal abstract fun readAllData(): Pair<ReadOnlyBytes, Long>

    /** A read-only view of the in-memory data backing this body, or `null` for file-backed bodies. */
    open val inMemoryData: ReadOnlyBytes? get() = null

    /** The file backing this body, or `null` for in-memory bodies. */
    open val file: File? get() = null

    /** The byte offset within the backing file where this body begins (0 for in-memory). */
    open val fileOffset: Long get() = 0

    /** The temp file owner, if any. Used to propagate ownership to derived bodies. */
    internal open val fileOwner: TempFileOwner? get() = null

    /**
     * A body backed by in-memory data.
     */
    class InMemory(val data: ByteArray) : RequestBody() {
        override val size: Long get() = data.size.toLong()
        override val inMemoryData: ReadOnlyBytes get() = ReadOnlyBytes(data)

        override fun inputStream(): InputStream =
            java.io.ByteArrayInputStream(data)

        override fun readBytes(): ByteArray = data.copyOf()

        override fun readAllData(): Pair<ReadOnlyBytes, Long> = ReadOnlyBytes(data) to 0L

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            return other is InMemory && data.contentEquals(other.data)
        }

        override fun hashCode(): Int = data.contentHashCode()
    }

    /**
     * A body backed by a byte range within a file on disk.
     *
     * Bytes are not read until [inputStream] or [readBytes] is called,
     * keeping the representation lightweight for uses like multipart part bodies.
     */
    class FileBacked internal constructor(
        override val file: File,
        override val fileOffset: Long = 0,
        override val size: Long,
        override val fileOwner: TempFileOwner? = null
    ) : RequestBody() {

        override fun inputStream(): InputStream {
            return object : InputStream() {
                private val raf = RandomAccessFile(file, "r").also { it.seek(fileOffset) }
                private var remaining = size
                private var closed = false

                override fun read(): Int {
                    if (remaining <= 0) return -1
                    remaining--
                    return raf.read()
                }

                override fun read(b: ByteArray, off: Int, len: Int): Int {
                    if (remaining <= 0) return -1
                    val toRead = minOf(len.toLong(), remaining).toInt()
                    val n = raf.read(b, off, toRead)
                    if (n > 0) remaining -= n
                    return n
                }

                override fun close() {
                    if (!closed) {
                        closed = true
                        raf.close()
                    }
                }
            }
        }

        override fun readBytes(): ByteArray {
            require(size <= Int.MAX_VALUE) { "Body too large to read into memory: $size bytes" }
            val intSize = size.toInt()
            RandomAccessFile(file, "r").use { raf ->
                raf.seek(fileOffset)
                val buf = ByteArray(intSize)
                var pos = 0
                while (pos < intSize) {
                    val n = raf.read(buf, pos, intSize - pos)
                    if (n == -1) break
                    pos += n
                }
                return if (pos == intSize) buf else buf.copyOf(pos)
            }
        }

        override fun readAllData(): Pair<ReadOnlyBytes, Long> = ReadOnlyBytes(readBytes()) to fileOffset

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            return other is FileBacked &&
                file == other.file &&
                fileOffset == other.fileOffset &&
                size == other.size
        }

        override fun hashCode(): Int {
            var result = file.hashCode()
            result = 31 * result + fileOffset.hashCode()
            result = 31 * result + size.hashCode()
            return result
        }
    }
}

/**
 * Abstraction over the parser's backing store.
 *
 * Tries to use a temp file on disk (suitable for large bodies). If the file
 * cannot be created, falls back to an in-memory buffer automatically.
 *
 * When memory-backed, the buffer is capped at [maxSize] to prevent unbounded
 * growth. The file-backed path has no cap — body size enforcement is handled
 * by the parser via `Content-Length` and `maxBodySize`.
 *
 * @param maxSize Maximum bytes allowed in the in-memory fallback buffer.
 *   Ignored when the buffer is file-backed.
 * @param cacheDir Optional directory for temp files (e.g., from `Context.getCacheDir()`).
 *   Files are created in a [tempSubdir] subdirectory. When `null`,
 *   falls back to the system temp directory.
 * @param tempSubdir The subdirectory name for temp files under [cacheDir].
 *   Defaults to [TempFileOwner.DEFAULT_TEMP_SUBDIR].
 */
internal class Buffer(maxSize: Int = Int.MAX_VALUE, cacheDir: File? = null, tempSubdir: String = TempFileOwner.DEFAULT_TEMP_SUBDIR) : Closeable {
    private val maxSize = maxSize
    private var file: File? = null
    private var raf: RandomAccessFile? = null
    private var memoryBuffer: java.io.ByteArrayOutputStream? = null
    private var fileOwnershipTransferred = false

    init {
        try {
            val tempDir = if (cacheDir != null) {
                File(cacheDir, tempSubdir).also { it.mkdirs() }
            } else {
                null
            }
            val tempFile = File.createTempFile("GutenbergKitHTTP-", null, tempDir)
            // No deleteOnExit() — on Android, apps are killed by the OS rather
            // than shut down gracefully, so the JVM shutdown hook never fires.
            // Cleanup is handled by TempFileOwner.close() and cleanOrphans().
            val handle = RandomAccessFile(tempFile, "rw")
            file = tempFile
            raf = handle
        } catch (_: Exception) {
            // Temp file unavailable — buffer in memory instead.
            memoryBuffer = java.io.ByteArrayOutputStream()
        }
    }

    /**
     * Appends data to the buffer.
     *
     * @return `true` if the data was accepted, `false` if the in-memory
     *   buffer would exceed its size limit.
     */
    fun append(data: ByteArray): Boolean {
        val r = raf
        if (r != null) {
            r.seek(r.length())
            r.write(data)
            return true
        } else {
            val buf = memoryBuffer!!
            if (buf.size() + data.size > maxSize) {
                return false
            }
            buf.write(data)
            return true
        }
    }

    fun read(offset: Int, maxLength: Int): ByteArray {
        require(offset >= 0) { "offset must be non-negative, was $offset" }
        require(maxLength >= 0) { "maxLength must be non-negative, was $maxLength" }
        if (maxLength == 0) return ByteArray(0)
        val r = raf
        if (r != null) {
            r.seek(offset.toLong())
            val buf = ByteArray(maxLength)
            var pos = 0
            while (pos < maxLength) {
                val n = r.read(buf, pos, maxLength - pos)
                if (n <= 0) break
                pos += n
            }
            return if (pos == maxLength) buf else buf.copyOf(pos)
        } else {
            val bytes = memoryBuffer!!.toByteArray()
            val start = offset.coerceAtMost(bytes.size)
            // Use Long arithmetic to avoid Int overflow when offset + maxLength > Int.MAX_VALUE
            // (possible with large media files like videos).
            val end = (offset.toLong() + maxLength.toLong()).coerceAtMost(bytes.size.toLong()).toInt()
            return bytes.copyOfRange(start, end)
        }
    }

    /**
     * Transfers ownership of the backing file to a [TempFileOwner].
     *
     * After this call, the buffer will no longer delete the file on [close].
     * Returns `null` if the buffer is memory-backed or ownership was already transferred.
     */
    fun transferFileOwnership(): Pair<File, TempFileOwner>? {
        val f = file ?: return null
        if (fileOwnershipTransferred) return null
        fileOwnershipTransferred = true
        return f to TempFileOwner(f)
    }

    /** Whether this buffer is backed by a file (vs. in-memory). */
    val isFileBacked: Boolean get() = raf != null

    override fun close() {
        try { raf?.close() } catch (_: Exception) {}
        if (!fileOwnershipTransferred) {
            file?.delete()
        }
    }
}

File: ios/Sources/GutenbergKitHTTP/RequestBody.swift
import Foundation

/// Process-wide registry of temp files currently backing an in-flight request.
///
/// ``HTTPServer/cleanOrphanedTempFiles(in:)`` runs a delete-all sweep of a
/// server's temp directory on start to reclaim files orphaned by a crash. Two
/// server instances that share a name share that directory (e.g. two editors
/// open at once, or one being torn down as another starts), so the sweep would
/// otherwise delete the other instance's live buffers. Registering a file here
/// while it is in use makes the sweep skip it; files not registered are crash
/// orphans (no live owner in this process) and are removed.
///
/// Keyed by file name (a unique UUID), which is stable however the directory is
/// later enumerated.
enum ActiveTempFiles {
    private static let lock = NSLock()
    // Guarded by `lock` on every access.
    nonisolated(unsafe) private static var names = Set<String>()

    static func register(_ name: String) { lock.withLock { _ = names.insert(name) } }
    static func unregister(_ name: String) { lock.withLock { _ = names.remove(name) } }
    static func contains(_ name: String) -> Bool { lock.withLock { names.contains(name) } }
}

/// Reference-counted owner for a temporary file.
///
/// The file is deleted when the last reference is released. This allows
/// ``RequestBody`` (a value type) to share ownership of a temp file across
/// copies — including multipart part bodies that reference byte ranges within
/// the same file. While owned, the file is registered in ``ActiveTempFiles`` so
/// a concurrent server's orphan sweep won't delete it.
final class TempFileOwner: Sendable {
    let url: URL
    init(url: URL) {
        self.url = url
        ActiveTempFiles.register(url.lastPathComponent)
    }
    deinit {
        ActiveTempFiles.unregister(url.lastPathComponent)
        try? FileManager.default.removeItem(at: url)
    }
}

/// An HTTP request body with stream semantics.
///
/// `RequestBody` abstracts over the underlying storage (in-memory data or a file on disk)
/// and provides uniform access regardless of backing:
/// - **Stream access**: Use ``makeInputStream()`` to read without loading everything into memory.
/// - **Materialized access**: Use ``data`` to get the full contents. For file-backed bodies,
///   this reads the entire file into memory.
public struct RequestBody: Sendable, Equatable {

    enum Storage: Sendable, Equatable {
        case data(Data)
        case file(URL)
        case fileSlice(url: URL, offset: UInt64, length: Int)
    }

    let storage: Storage

    /// Retains the backing file for parser-created bodies.
    /// Ignored in equality comparisons.
    private let _owner: TempFileOwner?

    public static func == (lhs: RequestBody, rhs: RequestBody) -> Bool {
        lhs.storage == rhs.storage
    }

    /// Creates a body backed by in-memory data.
    public init(data: Data) {
        self.storage = .data(data)
        self._owner = nil
    }

    /// Creates a body backed by a file on disk.
    ///
    /// The caller is responsible for ensuring the file exists for the lifetime of the body.
    public init(fileURL: URL) {
        self.storage = .file(fileURL)
        self._owner = nil
    }

    /// Creates a body backed by a byte range within a file on disk.
    ///
    /// The bytes are not read until ``makeInputStream()`` is called, keeping the
    /// representation lightweight for use cases like multipart part bodies.
    init(fileURL: URL, offset: UInt64, length: Int) {
        self.storage = .fileSlice(url: fileURL, offset: offset, length: length)
        self._owner = nil
    }

    /// Creates a body backed by an owned temporary file.
    ///
    /// The file is automatically deleted when the last `RequestBody` referencing it
    /// (including multipart part bodies derived from it) is released.
    init(ownedFileURL: URL) {
        self.storage = .file(ownedFileURL)
        self._owner = TempFileOwner(url: ownedFileURL)
    }

    /// Creates a body backed by a byte range within a file, sharing ownership
    /// with the source body's temp file.
    init(fileURL: URL, offset: UInt64, length: Int, owner: TempFileOwner?) {
        self.storage = .fileSlice(url: fileURL, offset: offset, length: length)
        self._owner = owner
    }

    /// The temp file owner, if any. Used to propagate ownership to derived bodies
    /// (e.g., multipart part slices).
    var fileOwner: TempFileOwner? { _owner }

    /// The number of bytes in the body.
    ///
    /// For in-memory and file-slice bodies this is O(1). For file-backed bodies
    /// this queries the file system without reading any data.
    public var count: Int {
        switch storage {
        case .data(let data):
            return data.count
        case .file(let url):
            return (try? url.resourceValues(forKeys: [.fileSizeKey]))?.fileSize ?? 0
        case .fileSlice(_, _, let length):
            return length
        }
    }

    /// The file URL backing this body, or `nil` for in-memory bodies.
    public var fileURL: URL? {
        switch storage {
        case .data: return nil
        case .file(let url), .fileSlice(let url, _, _): return url
        }
    }

    /// The byte offset within the backing file where this body begins.
    /// Returns 0 for in-memory and whole-file bodies.
    public var fileOffset: UInt64 {
        switch storage {
        case .fileSlice(_, let offset, _): return offset
        default: return 0
        }
    }

    /// The in-memory data backing this body, or `nil` for file-backed bodies.
    ///
    /// Unlike ``data``, this does **not** read from disk.
    public var inMemoryData: Data? {
        switch storage {
        case .data(let data): return data
        default: return nil
        }
    }

    /// The full body contents as `Data`.
    ///
    /// For in-memory bodies this returns the data directly. For file-backed
    /// bodies the file is read on a background thread to avoid blocking.
    ///
    /// You should almost always use `InputStream` instead.
    public var data: Data {
        get async throws {
            switch storage {
            case .data(let data):
                return data
            case .file(let url):
                return try Data(contentsOf: url)
            case .fileSlice(let url, let offset, let length):
                return try FileHandle.withReadHandle(forUrl: url) {
                    try $0.seek(toOffset: offset)
                    return try $0.read(upToCount: length) ?? Data()
                }
            }
        }
    }

    /// Creates an `InputStream` for reading the body contents.
    ///
    /// For file-slice bodies, a bound stream pair is used instead of subclassing
    /// `InputStream`. Subclassing `InputStream` with `super.init(data:)` triggers
    /// Foundation's class cluster design, causing `URLSession` to read from the
    /// empty superclass `Data` instead of the overridden `read(_:maxLength:)`.
    /// The bound stream pair avoids this because `Stream.getBoundStreams` returns
    /// a native `InputStream` that `URLSession` handles correctly.
    ///
    /// - Throws: A `CocoaError` if the backing file does not exist, is not readable, or is a directory.
    public func makeInputStream() throws -> InputStream {
        switch storage {
        case .data(let data):
            return InputStream(data: data)
        case .file(let url):
            return try Self.openFileStream(url: url)
        case .fileSlice(let url, let offset, let length):
            return try Self.makePipedFileSliceStream(url: url, offset: offset, length: length)
        }
    }

    /// Reads the entire body contents into memory and returns the data along with the
    /// file offset at which the data begins (0 for in-memory bodies).
    ///
    /// This is intended for scanning operations (e.g., multipart boundary detection)
    /// where the full body must be examined. The caller should release the returned
    /// `Data` as soon as scanning is complete.
    func readAllData() throws -> (Data, UInt64) {
        return switch storage {
        case .data(let data): (data, 0)
        case .file(let url): try (Data(contentsOf: url), 0)
        case .fileSlice(let url, let offset, let length):
            try FileHandle.withReadHandle(forUrl: url) {
                try $0.seek(toOffset: offset)
                return (try $0.read(upToCount: length) ?? Data(), offset)
            }
        }
    }

    /// Creates an `InputStream` backed by a bound stream pair that reads a byte
    /// range from a file on a background thread.
    ///
    /// The writer thread reads chunks from the file and pushes them into the
    /// `OutputStream` end of the pair. The returned `InputStream` is a native
    /// Foundation stream that `URLSession` and other consumers handle correctly.
    /// Backpressure is automatic: `OutputStream.write` blocks when the internal
    /// buffer is full.
    private static func makePipedFileSliceStream(url: URL, offset: UInt64, length: Int) throws -> InputStream {
        guard length > 0 else {
            return InputStream(data: Data())
        }

        let fileHandle = try FileHandle(forReadingFrom: url)
        try fileHandle.seek(toOffset: offset)

        var readStream: InputStream?
        var writeStream: OutputStream?
        Stream.getBoundStreams(withBufferSize: 65_536, inputStream: &readStream, outputStream: &writeStream)

        guard let inputStream = readStream, let outputStream = writeStream else {
            try? fileHandle.close()
            throw CocoaError(.fileReadUnknown, userInfo: [NSFilePathErrorKey: url.path])
        }

        outputStream.open()

        // OutputStream is not Sendable but is safely transferred to the
        // writer thread — only the thread accesses it after this point.
        nonisolated(unsafe) let output = outputStream

        Thread.detachNewThread {
            defer {
                output.close()
                try? fileHandle.close()
            }

            var remaining = length
            while remaining > 0 {
                let chunkSize = min(65_536, remaining)
                guard let chunk = try? fileHandle.read(upToCount: chunkSize),
                      !chunk.isEmpty else {
                    break
                }

                var written = 0
                chunk.withUnsafeBytes { buffer in
                    guard let base = buffer.baseAddress?.assumingMemoryBound(to: UInt8.self) else { return }
                    while written < chunk.count {
                        let result = output.write(base.advanced(by: written), maxLength: chunk.count - written)
                        if result <= 0 { return }
                        written += result
                    }
                }

                if written < chunk.count { break }
                remaining -= chunk.count
            }
        }

        return inputStream
    }

    private static func openFileStream(url: URL) throws -> InputStream {
        let path = url.path
        var isDirectory: ObjCBool = false

        guard FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory) else {
            throw CocoaError(.fileNoSuchFile, userInfo: [NSFilePathErrorKey: path])
        }

        guard !isDirectory.boolValue else {
            throw CocoaError(.fileReadInvalidFileName, userInfo: [NSFilePathErrorKey: path])
        }

        guard FileManager.default.isReadableFile(atPath: path) else {
            throw CocoaError(.fileReadNoPermission, userInfo: [NSFilePathErrorKey: path])
        }

        guard let stream = InputStream(url: url) else {
            throw CocoaError(.fileReadUnknown, userInfo: [NSFilePathErrorKey: path])
        }

        return stream
    }
}
